### PR TITLE
Generalize SerEnum

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
This PR is built on top of #2394.

### CODE REVIEW

**Comment by hesingh** [comment]
Would all the changes made to typeChecker.cpp be necessary if we merely added a preorder(IR::SerEnum) method which would use getTypeType()?

**Comment by mbudiu-vmw** [comment]
You can go ahead and try it. In preorder the type has not yet been computed. The types are computed bottom-up, so they are only available in postorder.

File: backends/ebpf/targets/xdp_target.py [hunk 1]
```diff
@@ -0,0 +1 @@
+/home/mbudiu/git/p4c/extensions/p4c-xdp/xdp_target.py
```
Comment (jnfoster) [inline backends/ebpf/targets/xdp_target.py:1]: This seems unrelated to the PR and also references a file in @mbudiu-vmw's machine...

**Comment by mbudiu-vmw** [comment]
@jnfoster : I think your change request has been satisfied.

**Comment by jnfoster** [comment]
Whoops. Sorry for letting this sit for 2 days. Thanks for the nudge.

**Comment by jafingerhut** [comment]
I do not know yet whether this will impact the P4Runtime API, but have created this issue to track the request for the P4 API working group to think about it: https://github.com/p4lang/p4runtime/issues/300

**Comment by smolkaj** [comment]
This PR got merged with a symlink to @mbudiu-vmw's machine, see @jnfoster's comment above.
